File: Pastura/Pastura/Models/TurnOutput.swift
import Foundation

/// Parsed output from a single LLM inference turn.
///
/// Wraps a `[String: String]` dictionary with typed accessors for common fields.
/// All values are normalized to `String` by `JSONResponseParser` regardless of
/// the original JSON type.
nonisolated public struct TurnOutput: Codable, Sendable, Equatable {
  /// The raw parsed fields from the LLM's JSON response.
  public let fields: [String: String]

  /// The unfiltered pre-parse LLM emission. Populated by
  /// ``JSONResponseParser/parse(_:)`` so that
  /// ``TurnRecord/rawOutput`` can store the original model output for
  /// audit (matching its documented contract). Absent (`nil`) for
  /// synthetic `TurnOutput` values constructed directly from `fields`
  /// in tests / replay paths.
  ///
  /// Excluded from `Codable` (see ``CodingKeys``) so the encoded shape
  /// used by ``TurnRecord/parsedOutputJSON`` stays stable and does not
  /// duplicate the ~1–2 KB of raw text already held in
  /// ``TurnRecord/rawOutput``. Ignored by `Equatable` because it is
  /// provenance metadata — two outputs with identical fields but
  /// different raw inputs represent the same domain value.
  public let rawText: String?

  public init(fields: [String: String], rawText: String? = nil) {
    self.fields = fields
    self.rawText = rawText
  }

  private enum CodingKeys: String, CodingKey {
    case fields
  }

  public init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    self.fields = try container.decode([String: String].self, forKey: .fields)
    self.rawText = nil
  }

  public func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encode(fields, forKey: .fields)
  }

  public static func == (lhs: TurnOutput, rhs: TurnOutput) -> Bool {
    lhs.fields == rhs.fields
  }

  // MARK: - Typed Accessors

  /// Agent's spoken statement (canonical primary field for speak phases).
  public var statement: String? { fields["statement"] }

  /// Agent's vote target name (canonical primary field for `.vote` phases).
  public var vote: String? { fields["vote"] }

  /// Agent's chosen action (canonical primary field for `.choose` phases —
  /// e.g., "cooperate" or "betray"). `OutputSchema.from(phase:)` binds this
  /// field to a GBNF enum constraint when the phase has non-empty `options`.
  public var action: String? { fields["action"] }

  /// Agent's private inner thought (hidden by default in UI, revealed on tap).
  public var innerThought: String? { fields["inner_thought"] }

  /// Reason for a vote or decision.
  public var reason: String? { fields["reason"] }

  // MARK: - Required Field Access

  /// Returns the value for the given key, or throws if the key is missing or empty.
  ///
  /// - Parameter key: The field key to look up.
  /// - Returns: The non-empty value for the key.
  /// - Throws: ``TurnOutputError/missingField(_:)`` if the key is absent or empty.
  public func require(_ key: String) throws -> String {
    guard let value = fields[key], !value.isEmpty else {
      throw TurnOutputError.missingField(key)
    }
    return value
  }

  /// Phase-aware extraction of the "primary" display text — the string
  /// that represents an agent's main visible action for the phase type.
  /// Used by UI and by streaming divergence telemetry to align a partial
  /// extractor snapshot with the canonical parse result.
  ///
  /// Speak / choose phases return the value at
  /// ``ScenarioConventions/primaryField(for:)`` (canonical-field lookup).
  /// `.vote` returns a composite `→ <voted> (<reason>)` string so the
  /// markdown export and live UI surface the reason inline. Code phases
  /// have no LLM output and return `nil`.
  public func primaryText(for phaseType: PhaseType) -> String? {
    if phaseType == .vote {
      return vote.map { voted in
        let reasonPart = reason.map { " (\($0))" } ?? ""
        return "→ \(voted)\(reasonPart)"
      }
    }
    guard let key = ScenarioConventions.primaryField(for: phaseType) else {
      return nil
    }
    return fields[key]
  }
}

/// Errors related to accessing ``TurnOutput`` fields.
nonisolated public enum TurnOutputError: Error, Sendable, Equatable {
  /// A required field was missing or empty in the LLM response.
  case missingField(String)
}

File: shared/models/src/commonMain/kotlin/com/pastura/models/TurnOutput.kt
package com.pastura.models

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

/**
 * Parsed output from a single LLM inference turn.
 *
 * Wraps a `Map<String, String>` with typed accessors for common fields.
 * All values are normalized to `String` by `JSONResponseParser` regardless of
 * the original JSON type.
 *
 * Kotlin port of `Pastura/Pastura/Models/TurnOutput.swift`.
 */
@Serializable
public data class TurnOutput(
    /** The raw parsed fields from the LLM's JSON response. */
    public val fields: Map<String, String>,
) {
    /**
     * The original pre-cleanup LLM text, kept so the Data layer can persist it in
     * the `TurnRecord.rawOutput` audit column. Populated by `JSONResponseParser`;
     * `null` for synthetic outputs built straight from [fields] (tests, replay).
     *
     * Declared in the class body rather than the primary constructor, so — exactly
     * like the Swift twin, which omits it from `CodingKeys` and from `==` — it is
     * excluded from [equals]/[hashCode] and from the serialized wire shape
     * (`@Transient`). Provenance metadata is not part of the domain value.
     *
     * Being outside the constructor also puts it **outside [copy]**: a `copy()`
     * returns an output with `rawText == null`, so re-set it explicitly whenever
     * a derived output must keep the provenance (see `WhisperHandler`).
     *
     * Written only before the owning event is emitted, never after — that
     * write-once discipline is what keeps the Swift side's
     * `@retroactive @unchecked Sendable` on `SimulationEvent` sound despite this
     * being a `var` (`.claude/rules/kmp-interop.md` Pattern 1).
     *
     * `public var` rather than a narrowed setter because `shared/engine`, which
     * writes it, is a separate Gradle module — `internal set` cannot reach it.
     */
    @Transient
    public var rawText: String? = null

    /** Agent's spoken statement (canonical primary field for speak phases). */
    public val statement: String? get() = fields["statement"]

    /** Agent's vote target name (canonical primary field for `.vote` phases). */
    public val vote: String? get() = fields["vote"]

    /**
     * Agent's chosen action (canonical primary field for `.choose` phases —
     * e.g., "cooperate" or "betray").
     */
    public val action: String? get() = fields["action"]

    /** Agent's private inner thought (hidden by default in UI, revealed on tap). */
    public val innerThought: String? get() = fields["inner_thought"]

    /** Reason for a vote or decision. */
    public val reason: String? get() = fields["reason"]

    /**
     * Returns the value for the given key, or throws if the key is missing or empty.
     *
     * `@Throws` is load-bearing at the K/N boundary, and the `@throws` line below
     * does not substitute for it (`.claude/rules/kmp-interop.md` Pattern 5).
     *
     * @param key The field key to look up.
     * @return The non-empty value for the key.
     * @throws [TurnOutputError.MissingField] if the key is absent or empty.
     */
    @Throws(TurnOutputError::class)
    public fun require(key: String): String {
        val value = fields[key]
        if (value.isNullOrEmpty()) throw TurnOutputError.MissingField(key)
        return value
    }

    /**
     * Phase-aware extraction of the "primary" display text — the string
     * that represents an agent's main visible action for the phase type.
     *
     * [PhaseType.VOTE] returns a composite `→ <voted> (<reason>)` string so the
     * markdown export and live UI surface the reason inline.
     * Speak / choose phases return the value at [ScenarioConventions.primaryField].
     * Code phases have no LLM output and return `null`.
     *
     * @param phaseType The phase type that produced this output.
     * @return The primary display text, or `null` for code phases or missing fields.
     */
    public fun primaryText(phaseType: PhaseType): String? {
        if (phaseType == PhaseType.VOTE) {
            val voted = vote ?: return null
            val reasonPart = reason?.let { " ($it)" } ?: ""
            return "→ $voted$reasonPart"
        }
        val key = ScenarioConventions.primaryField(phaseType) ?: return null
        return fields[key]
    }
}

/**
 * Errors related to accessing [TurnOutput] fields.
 *
 * Kotlin port of `Pastura/Pastura/Models/TurnOutput.swift:TurnOutputError`.
 * Extends [Exception] (Kotlin commonMain stdlib) for idiom alignment —
 * `RuntimeException` is not in commonMain.
 */
public sealed class TurnOutputError(message: String) : Exception(message) {
    /**
     * A required field was missing or empty in the LLM response.
     *
     * @property key The field key that was absent or empty.
     */
    public data class MissingField(public val key: String) : TurnOutputError("Missing field: $key")
}
